--- ./mm/vmscan.c	2021-07-17 21:24:05.854930547 -0700
+++ ./mm/vmscan.c	2021-07-17 21:34:44.151307895 -0700
@@ -973,9 +973,12 @@
 		if (reclaimed && page_is_file_cache(page) &&
 		    !mapping_exiting(mapping) && !dax_mapping(mapping))
 			shadow = workingset_eviction(page);
-		__delete_from_page_cache(page, shadow);
+       empty = __delete_from_page_cache(page, shadow);
 		xa_unlock_irqrestore(&mapping->i_pages, flags);
 
+       if (empty)
+           inode_pages_clear(mapping->host);
+
 		if (freepage != NULL)
 			freepage(page);
 	}
