--- ./tools/perf/arch/x86/entry/syscalls/syscall_64.tbl	2021-07-14 07:53:49.000000000 -0700
+++ ./tools/perf/arch/x86/entry/syscalls/syscall_64.tbl	2021-07-17 08:21:50.227994549 -0700
@@ -357,6 +357,9 @@
 433	common	fspick			__x64_sys_fspick
 434	common	pidfd_open		__x64_sys_pidfd_open
 435	common	clone3			__x64_sys_clone3/ptregs
+436    common  futex_wait      sys_futex_wait
+437    common  futex_wake      sys_futex_wake
+438    common  futex_waitv     sys_futex_waitv
 
 #
 # x32-specific system call numbers start at 512 to avoid cache impact
