--- linux-5.4.18/tools/perf/util/env.c	2020-02-05 13:22:53.000000000 -0800
+++ /home/arvin/Downloads/env.c	2020-02-09 12:22:11.109456896 -0800
@@ -258,13 +258,15 @@
 
 static int perf_env__read_arch(struct perf_env *env)
 {
-	struct utsname uts;
+	struct utsname *uts = NULL;
 
 	if (env->arch)
 		return 0;
 
-	if (!uname(&uts))
-		env->arch = strdup(uts.machine);
+	if (!uname(uts))
+		env->arch = strdup(uts->machine);
+    
+    free(uts);
 
 	return env->arch ? 0 : -ENOMEM;
 }
@@ -326,15 +328,17 @@
 
 const char *perf_env__arch(struct perf_env *env)
 {
-	struct utsname uts;
+	struct utsname *uts = NULL;
 	char *arch_name;
 
 	if (!env || !env->arch) { /* Assume local operation */
-		if (uname(&uts) < 0)
+		if (uname(uts) < 0)
 			return NULL;
-		arch_name = uts.machine;
+		arch_name = uts->machine;
 	} else
 		arch_name = env->arch;
 
+    free(uts);
+
 	return normalize_arch(arch_name);
 }
