--- ./arch/arm64/include/asm/unistd32.h	2021-07-11 03:52:08.000000000 -0700
+++ ./arch/arm64/include/asm/unistd32.h	2021-07-12 18:17:01.904234409 -0700
@@ -879,6 +879,8 @@
 __SYSCALL(__NR_pidfd_open, sys_pidfd_open)
 #define __NR_clone3 435
 __SYSCALL(__NR_clone3, sys_clone3)
+#define __NR_process_reap 436
+__SYSCALL(__NR_process_reap, sys_process_reap)
 
 /*
  * Please add new compat syscalls above this comment and update
