--- ./include/uapi/asm-generic/unistd.h	2021-07-11 03:52:08.000000000 -0700
+++ ./include/uapi/asm-generic/unistd.h	2021-07-12 18:21:13.484932552 -0700
@@ -850,9 +850,11 @@
 #define __NR_clone3 435
 __SYSCALL(__NR_clone3, sys_clone3)
 #endif
+#define __NR_process_reap 436
+__SYSCALL(__NR_process_reap, sys_process_reap)
 
 #undef __NR_syscalls
-#define __NR_syscalls 436
+#define __NR_syscalls 437
 
 /*
  * 32 bit systems traditionally used different
