--- ./kernel/sched/fair.c	2021-07-20 12:49:01.320759794 -0700
+++ ./kernel/sched/fair.c	2021-07-20 12:51:33.105114657 -0700
@@ -6716,6 +6716,7 @@
  */
 static void migrate_task_rq_fair(struct task_struct *p, int new_cpu)
 {
+#if !defined(CONFIG_CACULE_SCHED)
 	/*
 	 * As blocked tasks retain absolute vruntime the migration needs to
 	 * deal with this by subtracting the old and adding the new
@@ -6741,6 +6742,7 @@
 
 		se->vruntime -= min_vruntime;
 	}
+#endif /* CONFIG_CACULE_SCHED */
 
 	if (p->on_rq == TASK_ON_RQ_MIGRATING) {
 		/*
