--- ./kernel/sched/fair.c	2021-07-20 13:15:58.302280143 -0700
+++ ./kernel/sched/fair.c	2021-07-20 13:18:08.806204195 -0700
@@ -7370,10 +7370,10 @@
 	/*
 	 * Buddy candidates are cache hot:
 	 */
-	if (sched_feat(CACHE_HOT_BUDDY) && env->dst_rq->nr_running &&
-			(&p->se == cfs_rq_of(&p->se)->next ||
-			 &p->se == cfs_rq_of(&p->se)->last))
-		return 1;
+	//if (sched_feat(CACHE_HOT_BUDDY) && env->dst_rq->nr_running &&
+	//		(&p->se == cfs_rq_of(&p->se)->next ||
+	//		 &p->se == cfs_rq_of(&p->se)->last))
+	//	return 1;
 
 	if (sysctl_sched_migration_cost == -1)
 		return 1;
