--- ./drivers/gpu/drm/amd/display/modules/inc/mod_shared.h	2020-12-01 23:49:57.000000000 -0800
+++ ./drivers/gpu/drm/amd/display/modules/inc/mod_shared.h	2020-12-01 23:49:57.000000000 -0800
@@ -40,12 +40,12 @@
 
 enum vrr_packet_type {
 	PACKET_TYPE_VRR,
-	PACKET_TYPE_FS1,
-	PACKET_TYPE_FS2,
+	PACKET_TYPE_FS_V1,
+	PACKET_TYPE_FS_V2,
+	PACKET_TYPE_FS_V3,
 	PACKET_TYPE_VTEM
 };
 
-#if defined(CONFIG_DRM_AMD_DC_DCN2_0)
 union lut3d_control_flags {
 	unsigned int raw;
 	struct {
@@ -104,6 +104,5 @@
 	enum lut3d_control_gamut_map map2;
 	enum lut3d_control_rotation_mode rotation2;
 };
-#endif
 
 #endif /* MOD_SHARED_H_ */
