--- ./include/net/tcp.h	2021-07-17 23:02:54.619771192 -0700
+++ ./include/net/tcp.h	2021-07-17 23:08:05.602917627 -0700
@@ -1077,8 +1077,10 @@
 	u32  (*undo_cwnd)(struct sock *sk);
 	/* hook for packet ack accounting (optional) */
 	void (*pkts_acked)(struct sock *sk, const struct ack_sample *sample);
-	/* override sysctl_tcp_min_tso_segs */
-	u32 (*min_tso_segs)(struct sock *sk);
+   /* pick target number of segments per TSO/GSO skb (optional): */
+   u32 (*tso_segs)(struct sock *sk, unsigned int mss_now);
+   /* react to a specific lost skb (optional) */
+   void (*skb_marked_lost)(struct sock *sk, const struct sk_buff *skb);
 	/* returns the multiplier used in tcp_sndbuf_expand (optional) */
 	u32 (*sndbuf_expand)(struct sock *sk);
 	/* call when packets are delivered to update cwnd and pacing rate,
