--- ./kernel/futex.c	2021-07-10 20:26:43.743759674 -0700
+++ /home/arvin/Downloads/futex.c	2021-07-10 20:25:43.251244206 -0700
@@ -199,6 +199,8 @@
  * @bitset:		bitset for the optional bitmasked wakeup
  * @requeue_state:	State field for futex_requeue_pi()
  * @requeue_wait:	RCU wait for futex_requeue_pi() (RT only)
+ * @uaddr:             userspace address of futex
+ * @uval:              expected futex's value
  *
  * We use this hashed waitqueue, instead of a normal wait_queue_entry_t, so
  * we can wake only the relevant ones (hashed queues may be shared).
@@ -221,6 +223,8 @@
 	struct rt_mutex_waiter *rt_waiter;
 	union futex_key *requeue_pi_key;
 	u32 bitset;
+    u32 __user *uaddr;
+    u32 uval;
 	atomic_t requeue_state;
 #ifdef CONFIG_PREEMPT_RT
 	struct rcuwait requeue_wait;
