--- ./block/bfq-iosched.c	2021-07-14 20:17:21.356535211 -0700
+++ ./block/bfq-iosched.c	2021-07-14 20:20:19.867285424 -0700
@@ -2669,7 +2669,9 @@
 		if (bic->stable_merge_bfqq &&
 		    !bfq_bfqq_just_created(bfqq) &&
 		    time_is_after_jiffies(bfqq->split_time +
-					   msecs_to_jiffies(200))) {
+                      msecs_to_jiffies(200)) &&
+            time_is_before_jiffies(bfqq->creation_time +
+                       msecs_to_jiffies(200))) {
 			struct bfq_queue *stable_merge_bfqq =
 				bic->stable_merge_bfqq;
 			int proc_ref = min(bfqq_process_refs(bfqq),
