--- ./drivers/gpu/drm/amd/powerplay/hwmgr/smu7_hwmgr.c	2021-07-25 07:38:02.112348196 -0700
+++ ./drivers/gpu/drm/amd/powerplay/hwmgr/smu7_hwmgr.c	2021-07-25 07:39:57.200032230 -0700
@@ -36,6 +36,10 @@
 #include "hardwaremanager.h"
 #include "process_pptables_v1_0.h"
 #include "cgs_common.h"
+#include "pp_thermal.h"
+#include "smu7_baco.h"
+#include "smu7_smumgr.h"
+#include "polaris10_smumgr.h"
 
 #include "smu7_common.h"
 
