--- ./mm/compaction.c	2021-07-11 17:59:55.375229637 -0700
+++ ./mm/compaction.c	2021-07-11 17:59:43.858097825 -0700
@@ -1591,7 +1591,7 @@
  * Allow userspace to control policy on scanning the unevictable LRU for
  * compactable pages.
  */
-int sysctl_compact_unevictable_allowed __read_mostly = 1;
+int sysctl_compact_unevictable_allowed __read_mostly = 0;
 
 static inline void
 update_fast_start_pfn(struct compact_control *cc, unsigned long pfn)
