--- ./drivers/gpu/drm/amd/powerplay/hwmgr/smu7_hwmgr.c	2021-07-25 07:51:57.875271164 -0700
+++ ./drivers/gpu/drm/amd/powerplay/hwmgr/smu7_hwmgr.c	2021-07-25 07:53:28.914350150 -0700
@@ -4382,20 +4382,8 @@
 }
 
 static int
-smu7_notify_smc_display_change(struct pp_hwmgr *hwmgr, bool has_display)
-{
-	PPSMC_Msg msg = has_display ? (PPSMC_Msg)PPSMC_HasDisplay : (PPSMC_Msg)PPSMC_NoDisplay;
-
-	return (smum_send_msg_to_smc(hwmgr, msg) == 0) ?  0 : -1;
-}
-
-static int
 smu7_notify_smc_display_config_after_ps_adjustment(struct pp_hwmgr *hwmgr)
 {
-	if (hwmgr->display_config->num_display > 1 &&
-			!hwmgr->display_config->multi_monitor_in_sync)
-		smu7_notify_smc_display_change(hwmgr, false);
-
 	return 0;
 }
 
