--- ./mm/page-writeback.c	2021-07-25 00:47:02.256791476 -0700
+++ ./mm/page-writeback.c	2021-07-25 00:49:30.897051578 -0700
@@ -2813,8 +2813,12 @@
 						   PAGECACHE_TAG_WRITEBACK);
 
 			xas_set_mark(&xas, PAGECACHE_TAG_WRITEBACK);
-			if (bdi_cap_account_writeback(bdi))
-				inc_wb_stat(inode_to_wb(inode), WB_WRITEBACK);
+           if (bdi->capabilities & BDI_CAP_WRITEBACK_ACCT) {
+               struct bdi_writeback *wb = inode_to_wb(inode);
+               inc_wb_stat(wb, WB_WRITEBACK);
+               if (!on_wblist)
+                   wb_inode_writeback_start(wb);
+           }
 
 			/*
 			 * We can come through here when swapping anonymous
