--- ./block/bfq-iosched.c	2021-07-14 19:52:55.389935258 -0700
+++ ./block/bfq-iosched.c	2021-07-14 19:56:00.034204008 -0700
@@ -6567,7 +6567,7 @@
 		/* If the queue was seeky for too long, break it apart. */
 		if (bfq_bfqq_coop(bfqq) && bfq_bfqq_split_coop(bfqq) &&
 			!bic->stably_merged) {
-           bfq_log_bfqq(bfqd, bfqq, "breaking apart bfqq");
+           struct bfq_queue *old_bfqq = bfqq;
 
 			/* Update bic before losing reference to bfqq */
 			if (bfq_bfqq_in_large_burst(bfqq))
