--- ./include/net/tcp.h	2021-07-24 10:58:02.458546934 -0700
+++ ./include/net/tcp.h	2021-07-24 10:59:32.511376912 -0700
@@ -1013,6 +1013,12 @@
 #define TCP_CONG_NON_RESTRICTED 0x1
 /* Requires ECN/ECT set on all packets */
 #define TCP_CONG_NEEDS_ECN	0x2
+/* Wants notification of CE events (CA_EVENT_ECN_IS_CE, CA_EVENT_ECN_NO_CE). */
+#define TCP_CONG_WANTS_CE_EVENTS   0x4
+#define TCP_CONG_MASK  (TCP_CONG_NON_RESTRICTED | \
+        TCP_CONG_NEEDS_ECN | \
+        TCP_CONG_WANTS_CE_EVENTS)
+
 
 union tcp_cc_info;
 
