--- ./mm/workingset.c	2021-07-11 18:28:30.114074660 -0700
+++ ./mm/workingset.c	2021-07-11 18:34:36.071177036 -0700
@@ -222,7 +222,9 @@
 
 	lruvec = mem_cgroup_lruvec(pgdat, memcg);
 	eviction = atomic_long_inc_return(&lruvec->inactive_age);
-	return pack_shadow(memcgid, pgdat, eviction, PageWorkingset(page));
+   eviction >>= bucket_order; 
+   eviction = (eviction << WORKINGSET_WIDTH) | PageWorkingset(page);
+   return pack_shadow(memcgid, pgdat, eviction);
 }
 
 /**
@@ -267,9 +269,10 @@
 	memcg = mem_cgroup_from_id(memcgid);
 	if (!mem_cgroup_disabled() && !memcg)
 		goto out;
-	lruvec = mem_cgroup_lruvec(pgdat, memcg);
-	refault = atomic_long_read(&lruvec->inactive_age);
-	active_file = lruvec_lru_size(lruvec, LRU_ACTIVE_FILE, MAX_NR_ZONES);
+   eviction_lruvec = mem_cgroup_lruvec(eviction_memcg, pgdat);
+   refault = atomic_long_read(&eviction_lruvec->nonresident_age);
+   workingset = eviction & (BIT(WORKINGSET_WIDTH) - 1);
+   eviction = (eviction >> WORKINGSET_WIDTH) << bucket_order;
 
 	/*
 	 * Calculate the refault distance
