--- ./block/bfq-iosched.c	2021-07-14 16:41:33.634866521 -0700
+++ ./block/bfq-iosched.c	2021-07-14 16:41:39.470851916 -0700
@@ -2020,7 +2020,6 @@
 	if (!bfqd->last_completed_rq_bfqq ||
 	    bfqd->last_completed_rq_bfqq == bfqq ||
 	    bfq_bfqq_has_short_ttime(bfqq) ||
-	    bfqq->dispatched > 0 ||
 	    now_ns - bfqd->last_completion >= 4 * NSEC_PER_MSEC ||
 	    bfqd->last_completed_rq_bfqq == bfqq->waker_bfqq)
 		return;
